packages/contracts/contracts/abstracts/libs/logic/AtomicSwapStateLogic.sol

```diff
@@ -59,12 +62,444 @@ library AtomicSwapStateLogic {
+        if (selectedBid.expireTimestamp < block.timestamp) {
+            revert IAtomicSwapBase.BidAlreadyExpired(
+                selectedBid.expireTimestamp,
+                block.timestamp
+            );
+        }
+
+        // update the order status to 'COMPLETE'
+        order.status = IAtomicSwapBase.OrderStatus.COMPLETE;
+        order.completedAt = block.timestamp;
+        order.taker = bidder;
+
+        // Update the bid status to 'Executed'
+        selectedBid.status = IAtomicSwapBase.BidStatus.Executed;
+
+        AnteHandler.transferSellTokenToBuyer(
+            order,
+            swapOrderVestingParams[order.id], 
+            vestingManager,
+            selectedBid.bidder,
+            buyerFeeRate,
+            MAX_FEE_RATE_SCALE,
+            treasury
+        );
+        // Process buy token transfers
+        IAtomicSwapBase.Coin storage buyToken = order.buyToken;
+        AnteHandler.transferWithFee(
+            buyToken.token,
+            order.maker,
+            selectedBid.amount,
+            sellerFeeRate,
+            MAX_FEE_RATE_SCALE,
+            treasury
+        );
+        return (selectedBid.bidder, selectedBid.amount);
+    }
+
+       /// @notice Allows a bidder to cancel their bid on a swap order.
+    /// @param orderID The unique identifier of the swap order.
+    function cancelBid(
+        mapping( bytes32 => IAtomicSwapBase.AtomicSwapOrder ) storage swapOrder,
+        mapping( bytes32 => mapping( address => IAtomicSwapBase.Bid ) ) storage bids,
+        bytes32 orderID
+    ) external returns (address)  {
+        // Retrieve the selected bid from storage
+        IAtomicSwapBase.Bid storage selectedBid = bids[orderID][msg.sender];
+
+        // Ensure that msg.sender is same with bidder.
+        if (selectedBid.bidder != msg.sender) {
+            revert IAtomicSwapBase.UnauthorizedCancelAction();
+        }
+
+        // Ensure that the bid is in the 'Placed' status
+        if (selectedBid.status != IAtomicSwapBase.BidStatus.Placed) {
+            revert IAtomicSwapBase.BidNotInPlacedStatus(selectedBid.status);
+        }
+        // Update the bid status to 'Cancelled'
+        selectedBid.status = IAtomicSwapBase.BidStatus.Cancelled;
+
+        // Retrieve the buy tokens details from storage
+        IAtomicSwapBase.Coin storage buyToken = swapOrder[orderID].buyToken;
+
+        // If buy token is an ERC20 token, transfer it back to the bidder
+        if (buyToken.token != address(0)) {
+            TransferHelper.safeTransfer(
+                buyToken.token,
+                selectedBid.bidder,
+                selectedBid.amount
+            );
+        } else {
+            // If buy token is Ether, transfer it back to the bidder
+            TransferHelper.safeTransferETH(
+                msg.sender,
+                selectedBid.amount
+            );
+        }
+        return selectedBid.bidder;
+    }
+
+    function executeSwapWithPermit(
```

Comment:
the swap process looks good.  although there would not have asset lose if the swap failed due to insufficient balance in the implementation.  but we need provent spam transaction.